pkg/ha/ha_service.go

```diff
@@ -103,5 +105,70 @@ func (ha *HAService) Start(ctx context.Context) error {
 		}
 	}
 
-	return nil
+	<-ctx.Done()
+	err := ha.cleanUp()
+	if err == nil {
+		err = ctx.Err()
+	}
+	return err
+}
+
+// cleanUp is executed upon ending leadership. Its purpose is to remove the Endpoints object created upon acquiring
+// leadership.
+func (ha *HAService) cleanUp() error {
+	// Use our own context. This function executes when the main application context is closed.
+	// Also, try to finish before a potential 15 seconds termination grace timeout.
+	ctx, cancel := context.WithTimeout(context.Background(), 14*time.Second)
+	defer cancel()
+	seedClient := ha.manager.GetClient()
+
+	attempt := 0
+	var err error
+	for {
```

Comment:
I agree that it may have been better to use a poll function, but I don't think this it's worth refactoring. Replacing a "for" with which everybody is familiar, with а callback-based function in a domain-specific library, is a matter of preference.

Reply:
I'd personally prefer to use `Poll<...>` here unless there is a strong argument for the max number of attempts. Then I would stick with the for loop.
Generally, one benefit is that  `Poll<...>` should already be tested. Another is that when someone tries to do a similar wait and sees the `for{...}` loop he might decide to copy it and change it a bit, instead of simply reusing the `Poll<...>` function. As for domain-specificity - I think both GCMx and the functions in the https://github.com/kubernetes/kubernetes/blob/v1.28.0/staging/src/k8s.io/apimachinery/pkg/util/wait/poll.go share the same domain.

Reply:
@ialidzhikov @plkokanov 
Poll() is a better fit. No good argument against it. It was just that I didn't know if refactoring is worth it.

But the fact is that I haven't used Poll() before, so I can't really judge. I expect you're probably right and the improved readability justifies the  refactoring.

I'll use Poll().